src/basic/macro.h

```diff
@@ -463,4 +463,20 @@ typedef struct {
 
 assert_cc(sizeof(dummy_t) == 0);
 
+/* A little helper for subtracting 1 off a pointer in a safe UB-free way. This is intended to be used for for
+ * loops that count down from a high pointer until some base. A naive loop would implement this like this:
+ *
+ * for (p = end-1; p >= base; p--) …
+ *
+ * But this is not safe because p before the base is UB in C. With this macro the loop becomes this instead:
+ *
+ * for (p = PTR_SUB1(end, base); p; p = PTR_SUB1(p, base)) …
+ *
+ * And is free from UB! */
+#define PTR_SUB1(p, base)                                \
+        ({                                               \
+                typeof(p) _q = (p);                      \
+                _q && _q > (base) ? &_q[-1] : NULL;      \
```

Comment:
`&_q[-1]` works? O_O

Never seen this, imho explicit pointer arithmetic makes this easier to grok.

Reply:
> `&_q[-1]` works? O_O

what's worse: `1[q]` works too. Yes, you are reading right... In C `a[b] is just syntactic sugar for `*(a + b)`, and since additions are commutative, `a[b]` is identical to `b[a]`...

Fuck yeah, C FTW!

Reply:
the reason i didn't do pointer arith here, is to ensure PTR_SUB1() is only invokable on pointers/arrays and not on numeric values. Because one thing that C actually *does* is refuse the [] syntax if none of the involved arguments is a pointer/array...

Reply:
Wow. I just tested this and `1[q]` actually works. I am disgusted.